### DISCUSSION TITLE
Migrating old RCP application from PDE build

### DISCUSSION THREAD

**Discussion by sparklton** [eclipse-tycho/tycho#3778] [Q&A]
Hello everyone!
I am in the process to covert our legacy RCP application so that we can continue building it with Tycho as it seems that PDE build stopped working with the latest Eclipse. This is a released application with lots of users out there and we really don't want to break anything or cause disruptions with future p2 updates. Having read general Tycho docs and [tutorials ](https://www.vogella.com/tutorials/EclipseTycho/article.html#exercise-configure-your-modules), I am still at a loss how to get started. Perhaps a community can give me some guidance?

It's an Eclipse RCP application based on features. It' a pre-e4 application, so my first question is - will Tycho be able to build it?

This is the structure of the directories, - one bundle plugin with bunch of jars, the single feature, product definitions + main plugin. There are basically three folders (Eclipse projects) on the same level:

- bundle 
- feature 
- rcp application 

The application directory contains product definitions and main plugin - I am not sure this will be accepted by the Tycho as it doesn't follow the [recommended](https://tycho.eclipseprojects.io/doc/latest/StructuredBuild.html) structure. Do I just create a parent pom with three modules and Tycho will figure out how to build it all? Or should I restructure everything for Tycho to work?

**Comment by laeubi** [comment]
> It's an Eclipse RCP application based on features. It' a pre-e4 application, so my first question is - will Tycho be able to build it?

Tycho do not really care the type of application, so this will work.

> This is the structure of the directories, - one bundle plugin with bunch of jars, the single feature, product definitions + main plugin

If you follow Structured Build Layout then most of it should work seamless. As you correctly noticed the best would be to split the product file from the main plugin, so just create a dedicated plugin for it and place it into the `bundles` folder

In general that's what one can say base on this rather general information, if you can share a repository one might can give better details, beside that, if it is crucial to your business you can find dedicated support options for Tycho here:
https://github.com/eclipse-tycho/tycho#getting-support

**Reply by sparklton** [reply]
Thank you for the quick answer @laeubi!
I am going to give it a try, my primary concern was the pre-e4 dependencies and whether I can separate the main plugin from the product project..

**Comment by sparklton** [comment]
OK, I have managed to move on with Tycho build but stuck with the following error when doing `mvn clean verify`:

`Missing requirement: com.client.feature.feature.group 5.3.0 requires 'org.eclipse.equinox.p2.iu; org.eclipse.equinox.launcher.cocoa.macosx.aarch64 0.0.0' but it could not be found`

I have used the target platform file we use with the PDE, it includes everything from [this update site](https://download.eclipse.org/releases/2023-03/) for multiple platforms - windows, linux, macos (x86 and aarch).

I don't quite understand where the `org.eclipse.equinox.p2.iu` is coming from and why there is a problem to resolve  `org.eclipse.equinox.launcher.cocoa.macosx.aarch64` which is a part of the update site where target platform comes from.  

Appreciate any help!
